File: lib/features/rooms/models/room_model.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:popcorn/core/models/user_model.dart';
import 'package:flutter/foundation.dart';

part 'room_model.freezed.dart';
part 'room_model.g.dart';
@freezed
class RoomModel with _$RoomModel {

  const factory RoomModel.create({
    @JsonKey(name: '_id') required String id,
    required String author,
    required String name,
    required String description,
    String? bio,
    required int status,
    String? token,
    String? roomImage,
    String? backgroundImage,
    String? toastMessage,
    required List<String> members,
    required List<String> banned,
    required DateTime createdAt,
    required DateTime updatedAt,
  }) = RoomCreateModel;


  @JsonSerializable(explicitToJson: true)
  const factory RoomModel.author({
    @JsonKey(name: '_id') required String id,
    required UserModel author,
    required String name,
    required String description,
    String? bio,
    required int status,
    String? token,
    String? roomImage,
    String? backgroundImage,
    String? toastMessage,
    required List<String> members,
    required List<String> banned,
    required DateTime createdAt,
    required DateTime updatedAt,
  }) = RoomAuthorModel;


  factory RoomModel.fromJson(Map<String, Object?> json)
  => _$RoomModelFromJson(json);
}

File: functions/database/models/room_model.js
const mongoose = require('mongoose');
const Schema = mongoose.Schema;



const roomSchema = Schema(
    {
        '_id': {
            type: String,
            required: true,
        },
        'author': {
            type: String,
            ref: 'User',
            required: true,
        },
        'name': {
            type: String,
            required: true,
        },
        'description': {
            type: String,
            required: false,
        },
        'bio': {
            type: String,
            required: false,
        },
        'status': {
            type: Number,
            required: true,
        },
        'token': {
            type: String,
            required: false,
        },
        'roomImage': {
            type: String,
            required: false,
        },
        'backgroundImage': {
            type: String,
            required: false,
        },
        'toastMessage': {
            type: String,
            required: false,
        },
        'members': [{
            type: String,
            ref: 'User'
        }],
        'banned': [{
            type: String,
            ref: 'User'
        }],
        'createdAt': {
            type: Date,
            default: Date.now,
        },
        'updatedAt': {
            type: Date,
            default: Date.now,
        },
        'runtimeType': {
            type: String,
            required: false,
        },
    },
    {
        timestamps: true ,
        collection: 'rooms',
        toObject: {
            transform: function (doc, ret, options) {
                delete ret.__v;
            },
        },
    },
);

module.exports = mongoose.model('Room', roomSchema);
